update/v1alpha1/types_update_status.go

```diff
@@ -0,0 +1,750 @@
+type ClusterVersionStatusInsightUpdatingReason string
+
+const (
+	// CannotDetermineUpdating is used with Updating=Unknown
+	ClusterVersionCannotDetermineUpdating ClusterVersionStatusInsightUpdatingReason = "CannotDetermineUpdating"
+	// ClusterVersionProgressing means that ClusterVersion is considered to be Updating=True because it has a Progressing=True condition
+	ClusterVersionProgressing ClusterVersionStatusInsightUpdatingReason = "ClusterVersionProgressing"
+	// ClusterVersionNotProgressing means that ClusterVersion is considered to be Updating=False because it has a Progressing=False condition
+	ClusterVersionNotProgressing ClusterVersionStatusInsightUpdatingReason = "ClusterVersionNotProgressing"
+)
+
+// VersionMetadataType is the type of the metadata value
+// +kubebuilder:validation:Enum=string;bool;int
+type VersionMetadataType string
+
+const (
+	StringVersionMetadata VersionMetadataType = "string"
+	BoolVersionMetadata   VersionMetadataType = "bool"
+	IntVersionMetadata    VersionMetadataType = "int"
+)
+
+type VersionMetadataKey string
+
+const (
+	// installation denotes a boolean that indicates the update was initiated as an installation
+	InstallationMetadata VersionMetadataKey = "installation"
+	// partial denotes a boolean that indicates the update was initiated in a state where the previous upgrade
+	// (to the original version) was not fully completed
+	PartialMetadata VersionMetadataKey = "partial"
+	// architecture denotes a string that indicates the architecture of the payload image of the version,
+	// when relevant
+	ArchitectureMetadata VersionMetadataKey = "architecture"
+)
+
+type VersionMetadata struct {
+	// +required
+	Key VersionMetadataKey `json:"key"`
+
+	// +required
+	VersionMetadataValue `json:",inline"`
+}
+
+type VersionMetadataValue struct {
+	// +unionDiscriminator
+	// +required
+	// +kubebuilder:validation:Enum:string;bool;int
+	Type VersionMetadataType `json:"type"`
+
+	// +optional
+	// +kubebuilder:validation:Type=string
+	// +unionMember
+	String string `json:"string,omitempty"`
+
+	// +optional
+	// +kubebuilder:validation:Type=bool
+	// +unionMember
+	Bool bool `json:"bool,omitempty"`
+
+	// +optional
+	// +kubebuilder:validation:Type=int
+	// +unionMember
+	Integer int `json:"integer,omitempty"`
+}
+
+// Version describes a version involved in an update, typically on one side of an update edge
+type Version struct {
+	// version is a semantic version string
+	// +required
+	Version string `json:"version,omitempty"`
+
+	// metadata is a list of metadata associated with the version
+	// +listType=map
+	// +listMapKey=key
+	// +optional
+	Metadata []VersionMetadata `json:"metadata,omitempty"`
+}
+
+// ControlPlaneUpdateVersions contains the original and target versions of the upgrade
+type ControlPlaneUpdateVersions struct {
+	// previous is the version of the control plane before the update
+	// +optional
+	Previous Version `json:"previous,omitempty"`
```

Comment:
nit: `omitempty` doesn't do anything for structs (golang/go#11939, golang/go#51261).  You should either drop `omitempty` here, or use `*Version`.

Reply:
good point, thanks! I audited the rest of the file for this, and either made the field required/not-omitempty, or made it a pointer.

I think there was a guideline about not using pointers in one of the conventions, but [kube allows it](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required) and can't find [anything for openshift](https://github.com/openshift/enhancements/blob/master/CONVENTIONS.md#api).

Reply:
> I think there was a guideline about not using pointers in one of the conventions...

The OpenShift guideline for that is tucked away in [this policy doc][1].  There's a subsequent section allowing pointers to structs when [an API validation that requires the field to be unset][2], which doesn't seem to apply here.  I'm just not sold on "inline zero values to help discoverability" vs. "use `omitempty` to hide things the admin doesn't care about; future admins can learn about new knobs via `oc explain ...`".  Feel free to drop the pointers and the  `omitempty` if the API approvers press you, and feel free to mark this thread resolved either way.

[1]: https://github.com/openshift/enhancements/blob/7d163511f45e73e707f8ec17f34a446772b2cff6/dev-guide/api-conventions.md#pointers
[2]: https://github.com/openshift/enhancements/blob/7d163511f45e73e707f8ec17f34a446772b2cff6/dev-guide/api-conventions.md#pointers-to-structs

Reply:
Yeah, I eventually found  this guideline (the API conventions are linked from the top level conventions). The OpenShift guideline about avoiding pointers is about configuration APIs (_"In configuration APIs specifically,..."_) which Status API is not. The point about pointers being harder to work with still stands.

I ended  up using the pointer at a single place in the whole API, the `PoolResource` in `ControlPlane`. I reworked the `ControlPlaneUpdateVersions` (https://github.com/openshift/api/pull/2012/files#diff-382a9e9c3b33f1db2b3e759e6d6019ab4db0f3ea29c8448a281bcee9bc0f4c81R71-R81) part (also simplified the metadata, the type discriminated union felt like overkill etc) and now it is no longer a pointer.